File: android/app/src/main/kotlin/dev/omnesis/android/ui/common/AnnotationsSection.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.common

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.AutoAwesome
import androidx.compose.material3.Icon
import androidx.compose.material3.TextButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.omnesis.android.designsystem.theme.OmSpacing
import dev.omnesis.android.designsystem.theme.OmTheme
import dev.omnesis.android.transport.dto.Annotation
import dev.omnesis.android.transport.dto.AnnotationDependent
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.OffsetDateTime
import kotlin.math.roundToInt

/**
 * The agent's durable LLM-derived observations, rendered as a flat section. Person
 * annotations (about a person) and document annotations (grounded on a document) share the
 * same wire shape — claimType / claimText / confidence / evidenceQuote — so one section
 * serves both surfaces (the person detail + the document inspector). [title] is a full
 * sentence supplied by the caller ("Enriched by Omnesis", "Profile", or "What Omnesis has
 * learned about <name>"). Confidence is surfaced verbatim: these are defeasible
 * observations, not hard facts. Renders nothing when [annotations] is empty and
 * there is no pending page; an empty cursor-bearing page keeps only its sentinel
 * so pagination can make progress.
 *
 * The header is bespoke (sparkle + sentence-case title) rather than the shared
 * uppercase-rule `PeopleSectionHeader`/`FlatSectionHeader`: uppercasing a sentence title
 * like "What Omnesis has learned about Maya Reeves" mangles it. Matches the iOS
 * `AnnotationsSection` and the portal `AnnotationList` so the three clients read alike.
 */
fun LazyListScope.annotationsSectionItems(
    listState: LazyListState,
    keyPrefix: String,
    title: String,
    annotations: List<Annotation>,
    paging: CursorPagingState = CursorPagingState(),
    onLoadMore: () -> Unit = {},
    dependents: Map<String, AnnotationDependentsUi> = emptyMap(),
    onToggleDependents: (Annotation) -> Unit = {},
    onLoadMoreDependents: (Annotation) -> Unit = {},
) {
    if (
        annotations.isEmpty() &&
        !paging.canLoadMore &&
        !paging.isLoadingMore &&
        paging.paginationError == null &&
        paging.refreshError == null
    ) {
        return
    }
    if (annotations.isNotEmpty()) {
        item("$keyPrefix:header") {
            AnnotationsHeader(title)
        }
        annotations.forEachIndexed { index, annotation ->
            item("$keyPrefix:annotation:${annotation.id}") {
                AnnotationRow(
                    annotation = annotation,
                    dependents = dependents[annotation.id],
                    onToggleDependents = { onToggleDependents(annotation) },
                )
            }
            val dependentState = dependents[annotation.id]
            if (dependentState?.expanded == true) {
                items(
                    dependentState.items,
                    key = { "$keyPrefix:dependent:${annotation.id}:${it.kind}:${it.id}" },
                ) { dependent ->
                    AnnotationDependentRow(dependent)
                }
                item("$keyPrefix:dependents-footer:${annotation.id}") {
                    val boundaryKey = "$keyPrefix:dependents-footer:${annotation.id}"
                    if (dependentState.shouldRetryFirstPage) {
                        AutomaticPagingBoundary(
                            listState = listState,
                            boundaryKey = boundaryKey,
                            loadToken = null,
                            isLoading = false,
                            errorMessage = "Couldn't load uses.",
                            retryLabel = "Retry loading uses",
                            onLoadMore = { onLoadMoreDependents(annotation) },
                        )
                    } else {
                        ListPagingFooter(
                            listState = listState,
                            boundaryKey = boundaryKey,
                            paging = dependentState.paging,
                            onLoadMore = { onLoadMoreDependents(annotation) },
                            loadAction = "more uses",
                        )
                    }
                }
            }
            if (index < annotations.lastIndex) {
                item("$keyPrefix:divider:${annotation.id}") {
                    Box(
                        Modifier
                            .fillMaxWidth()
                            .height(1.dp)
                            .background(OmTheme.colors.borderLight),
                    )
                }
            }
        }
    }
    item("$keyPrefix:paging") {
        ListPagingFooter(
            listState = listState,
            boundaryKey = "$keyPrefix:paging",
            paging = paging,
            onLoadMore = onLoadMore,
            loadAction = "more observations",
        )
    }
}

data class AnnotationDependentsUi(
    val expanded: Boolean = false,
    val loading: Boolean = false,
    val items: List<AnnotationDependent> = emptyList(),
    val paging: CursorPagingState = CursorPagingState(),
) {
    val shouldRetryFirstPage: Boolean
        get() = items.isEmpty() && paging.refreshError != null
}

/**
 * Accent-tinted sparkle marks the section as Omnesis-derived (agent) data, distinct from
 * source-provided fields; the sentence-case title is kept verbatim (no uppercase, no rule).
 */
@Composable
private fun AnnotationsHeader(title: String) {
    val c = OmTheme.colors
    Row(
        modifier = Modifier.fillMaxWidth().padding(top = OmSpacing.md),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Icon(
            Icons.Outlined.AutoAwesome,
            contentDescription = null,
            tint = c.accent,
            modifier = Modifier.size(13.dp),
        )
        Text(
            title,
            fontSize = 12.sp,
            fontWeight = FontWeight.SemiBold,
            color = c.textSecondary,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
internal fun AnnotationRow(
    annotation: Annotation,
    dependents: AnnotationDependentsUi?,
    onToggleDependents: () -> Unit,
) {
    val c = OmTheme.colors
    Column(
        modifier = Modifier.fillMaxWidth().padding(vertical = OmSpacing.sm),
        verticalArrangement = Arrangement.spacedBy(OmSpacing.xs),
    ) {
        // Claim-type label + basis chip on the left; the verification marker
        // and the confidence right-aligned. The left group is a weighted inner
        // row so a long claim type still ellipsizes instead of pushing the
        // tail off screen.
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Row(
                Modifier.weight(1f),
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    annotation.claimType,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = c.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f, fill = false),
                )
                val basis = annotation.claimBasis
                if (!basis.isNullOrEmpty()) {
                    Text(
                        basis,
                        fontSize = 10.sp,
                        color = c.textMuted,
                        maxLines = 1,
                        modifier = Modifier
                            .background(c.bgTertiary, RoundedCornerShape(50))
                            .padding(horizontal = 6.dp, vertical = 1.dp),
                    )
                }
            }
            val verification = verificationLabel(annotation)
            if (verification != null) {
                Text(
                    verification,
                    fontSize = 10.sp,
                    color = c.textMuted,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            Text(
                "${(annotation.confidence * 100).roundToInt()}%",
                fontSize = 11.sp,
                color = c.textMuted,
                style = TextStyle(fontFeatureSettings = "tnum"),
            )
        }
        Text(annotation.claimText, fontSize = 13.sp, color = c.textPrimary)
        val quote = annotation.evidenceQuote
        if (quote != null && quote.isNotEmpty()) {
            Text(
                "“$quote”",
                fontSize = 12.sp,
                fontStyle = FontStyle.Italic,
                color = c.textMuted,
            )
        }
        if (annotation.dependentCount > 0) {
            TextButton(
                onClick = onToggleDependents,
                enabled = dependents?.loading != true,
                modifier = Modifier.padding(horizontal = 0.dp),
            ) {
                if (dependents?.loading == true) {
                    dev.omnesis.android.designsystem.components.OmSpinner(Modifier.size(14.dp))
                    Spacer(Modifier.width(OmSpacing.xs))
                    Text("Loading uses…")
                } else {
                    Text(
                        if (dependents?.expanded == true) {
                            "Hide where this was used"
                        } else {
                            "Used by ${annotation.dependentCount} " +
                                if (annotation.dependentCount == 1) "output" else "outputs"
                        },
                    )
                }
            }
        }
    }
}

@Composable
private fun AnnotationDependentRow(dependent: AnnotationDependent) {
    val c = OmTheme.colors
    Text(
        "${dependent.kind.uppercase()} · ${dependent.title.ifBlank { dependent.id }}",
        fontSize = 11.sp,
        color = c.textSecondary,
        maxLines = 2,
        overflow = TextOverflow.Ellipsis,
        modifier = Modifier
            .fillMaxWidth()
            .background(c.bgTertiary, RoundedCornerShape(OmTheme.radius.small))
            .padding(horizontal = OmSpacing.sm, vertical = OmSpacing.xs),
    )
}

/**
 * Entailment-check marker: "verified · 2d ago" when the last check time is
 * known, the bare state otherwise, null when the row carries no stamp.
 */
private fun verificationLabel(annotation: Annotation): String? {
    val state = annotation.verificationState?.takeIf { it.isNotEmpty() } ?: return null
    val checkedAt = annotation.lastVerifiedAt?.let(::parseIsoToEpochMillis) ?: return state
    // Beyond the relative window, keep the marker SHORT (a bare date) — the
    // portal and iOS render a short date for old stamps, and the full
    // date-time form would crowd the claim type off a narrow row.
    val relative = TimeFormat.relative(checkedAt)
    val label = if (relative.length > 12) shortDate(checkedAt) else relative
    return "$state · $label"
}

private fun shortDate(epochMillis: Long): String =
    DateTimeFormatter.ofPattern("d MMM")
        .withZone(ZoneId.systemDefault())
        .format(Instant.ofEpochMilli(epochMillis))

/** Lenient ISO-8601 parse (instant or offset form); null when unparseable. */
private fun parseIsoToEpochMillis(iso: String): Long? =
    runCatching { Instant.parse(iso).toEpochMilli() }
        .recoverCatching { OffsetDateTime.parse(iso).toInstant().toEpochMilli() }
        .getOrNull()

File: ios/Sources/Omnesis/UI/AnnotationsSection.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI

/// The agent's durable LLM annotations for a document or person, rendered as
/// a flat section. Person annotations (about a person) and document
/// annotations (grounded on a document) share the same wire shape
/// (`Annotation`), so one view serves both — the caller supplies the
/// sentence `title` ("Enriched by Omnesis", "Profile", or "What Omnesis has
/// learned about <name>"). Mirrors the portal's `AnnotationList`.
///
/// Confidence is surfaced verbatim: these are defeasible observations, not
/// hard facts. Renders nothing only when `annotations` is definitively empty;
/// paging progress, retry, and truncation states remain visible. v1 is
/// read-only text — there is no deep-link into the evidence document.
@available(iOS 17.0, *)
struct AnnotationsSection: View {
    let title: String
    let annotations: [Annotation]
    var paging = CursorPagingState()
    var onLoadMore: () -> Void = {}

    var body: some View {
        if shouldShowPagedContent(itemCount: annotations.count, paging: paging) {
            VStack(alignment: .leading, spacing: Theme.Spacing.sm) {
                header
                ForEach(Array(annotations.enumerated()), id: \.element.id) { idx, annotation in
                    row(annotation)
                    if idx < annotations.count - 1 {
                        Divider().background(Theme.borderLight)
                    }
                }
                ListPagingFooter(
                    state: paging,
                    label: "Load more annotations",
                    retry: onLoadMore
                )
            }
        }
    }

    /// A bespoke header rather than `FlatSectionHeader`: the latter
    /// force-uppercases its title and appends a full-width rule, which mangles
    /// a sentence title like "What Omnesis has learned about Maya Reeves". The
    /// sparkle marks the section as Omnesis-derived (agent) data, distinct
    /// from source-provided fields.
    private var header: some View {
        HStack(spacing: 6) {
            Image(systemName: "sparkles")
                .font(.system(size: 12))
                .foregroundStyle(Theme.accent)
            Text(title)
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(Theme.textSecondary)
                .lineLimit(2)
        }
        .padding(.top, Theme.Spacing.md)
    }

    private func row(_ a: Annotation) -> some View {
        VStack(alignment: .leading, spacing: Theme.Spacing.xs) {
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Text(a.claimType)
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.textPrimary)
                if let basis = a.claimBasis, !basis.isEmpty {
                    Text(basis)
                        .font(.system(size: 10))
                        .foregroundStyle(Theme.textMuted)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 1)
                        .background(Theme.bgTertiary)
                        .clipShape(Capsule())
                }
                Spacer(minLength: 6)
                if let verification = verificationLabel(a) {
                    Text(verification)
                        .font(.system(size: 10))
                        .foregroundStyle(Theme.textMuted)
                }
                Text("\(Int((a.confidence * 100).rounded()))%")
                    .font(.system(size: 11))
                    .foregroundStyle(Theme.textMuted)
                    .monospacedDigit()
            }
            Text(a.claimText)
                .font(.system(size: 13))
                .foregroundStyle(Theme.textPrimary)
                .fixedSize(horizontal: false, vertical: true)
            if let quote = a.evidenceQuote, !quote.isEmpty {
                Text("“\(quote)”")
                    .font(.system(size: 12))
                    .italic()
                    .foregroundStyle(Theme.textMuted)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.vertical, Theme.Spacing.sm)
    }

    /// Entailment-check marker: "verified · 2d ago" when the last check time
    /// is known, the bare state otherwise, nil when the row carries no stamp.
    private func verificationLabel(_ a: Annotation) -> String? {
        guard let state = a.verificationState, !state.isEmpty else { return nil }
        if let ago = formatTimeAgo(a.lastVerifiedAt) {
            return "\(state) · \(ago)"
        }
        return state
    }
}

#if DEBUG
@available(iOS 17.0, *)
private func annotationsPreview(title: String, annotations: [Annotation]) -> some View {
    ScrollView {
        AnnotationsSection(title: title, annotations: annotations)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(Theme.Spacing.lg)
    }
    .background(Theme.bgPrimary.ignoresSafeArea())
    .preferredColorScheme(.dark)
}

@available(iOS 17.0, *)
#Preview("Annotations — document (Enriched by Omnesis)") {
    annotationsPreview(title: "Enriched by Omnesis", annotations: PreviewMocks.documentAnnotations)
}

@available(iOS 17.0, *)
#Preview("Annotations — person self (Profile)") {
    annotationsPreview(title: "Profile", annotations: PreviewMocks.personAnnotations)
}

@available(iOS 17.0, *)
#Preview("Annotations — person other (learned about)") {
    annotationsPreview(
        title: "What Omnesis has learned about Maximilian Aurelius Bartholomew Hawthorne-Featherstonehaugh",
        annotations: PreviewMocks.personAnnotations
    )
}
#endif
#endif
